lib/recursive-methods.rb

```diff
@@ -1,49 +1,85 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
+require 'pry'
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  end
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
+# I think this is reversing the string in place?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return reverse(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return reverse_inplace(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return n if n == 0
+  return 2 + bunny(n-1)
 end
 
 # Time complexity: ?
 # Space complexity: ?
+# NOT PASSING TESTS
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length == 0
+  # need to check if s = even number of paren-pairs()
+  # return true if s.length % 2 == 0
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, i = 0)
+  return false if i >= array.length
+  return true if array[i] == value
+  return search(array, value, i + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+# ONLY PASSING 2 TESTS
+def is_palindrome(s, l = 0, r = s.length-1 )
+  if s[l] == s[r]
+    return true
+  else
+    return false
+  end
+
+  while l < s.length
```

Comment:
You don't need a while if it's recursive.

You can make the base case 

`return true if i >= r`